From 8465c40976260ebe72cc9dc32610d6d2b003e920 Mon Sep 17 00:00:00 2001
From: Aya ElAttar <ayaelattar@chromium.org>
Date: Tue, 8 Mar 2022 08:16:30 +0000
Subject: [PATCH] [Merge M99] Change DropHelper::OnDrop to call GetDropCallback

DropHelper::OnDrop is still used in DragDropClientMac,
and it calls OnPerformDrop which caused the regression
because OnPerformDrop is being removed. So I changed
OnDrop to use GetDropCallback instead.

(cherry picked from commit 8fc99d683768fbf34cb0beb43ee761fd50f42d68)

Bug: 1292896
Change-Id: I908cbecfaa6ffb336c0afc4581b2c3077aa8c47b
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3433146
Reviewed-by: Elly Fong-Jones <ellyjones@chromium.org>
Commit-Queue: Aya Elsayed <ayaelattar@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#967837}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3507992
Cr-Commit-Position: refs/branch-heads/4844@{#1007}
Cr-Branched-From: 007241ce2e6c8e5a7b306cc36c730cd07cd38825-refs/heads/main@{#961656}
---
 .../cocoa/drag_drop_client_mac_unittest.mm    | 25 ++++++++++++++++++-
 ui/views/widget/drop_helper.cc                |  5 +++-
 2 files changed, 28 insertions(+), 2 deletions(-)

--- a/ui/views/cocoa/drag_drop_client_mac_unittest.mm
+++ b/ui/views/cocoa/drag_drop_client_mac_unittest.mm
@@ -149,6 +149,14 @@
     return DragOperation::kMove;
   }
 
+  views::View::DropCallback GetDropCallback(
+      const ui::DropTargetEvent& event) override {
+    return base::BindOnce([](const ui::DropTargetEvent& event,
+                             ui::mojom::DragOperation& output_drag_op) {
+      output_drag_op = DragOperation::kMove;
+    });
+  }
+
  private:
   // Drop formats accepted by this View object.
   int formats_ = 0;
@@ -317,8 +325,23 @@
 
   // View:
   DragOperation OnPerformDrop(const ui::DropTargetEvent& event) override {
+    ui::mojom::DragOperation output_drag_op = ui::mojom::DragOperation::kNone;
+    PerformDrop(event, output_drag_op);
+    return output_drag_op;
+  }
+  views::View::DropCallback GetDropCallback(
+      const ui::DropTargetEvent& event) override {
+    // base::Unretained is safe here because in the tests the view isn't deleted
+    // before the drop callback is run.
+    return base::BindOnce(&DragDropCloseView::PerformDrop,
+                          base::Unretained(this));
+  }
+
+ private:
+  void PerformDrop(const ui::DropTargetEvent& event,
+                   ui::mojom::DragOperation& output_drag_op) {
     GetWidget()->CloseNow();
-    return DragOperation::kMove;
+    output_drag_op = DragOperation::kMove;
   }
 };
 
--- a/ui/views/widget/drop_helper.cc
+++ b/ui/views/widget/drop_helper.cc
@@ -101,7 +101,10 @@
   View::ConvertPointToTarget(root_view, drop_view, &view_location);
   ui::DropTargetEvent drop_event(data, gfx::PointF(view_location),
                                  gfx::PointF(view_location), drag_operation);
-  return drop_view->OnPerformDrop(drop_event);
+  auto output_drag_op = ui::mojom::DragOperation::kNone;
+  auto drop_cb = drop_view->GetDropCallback(drop_event);
+  std::move(drop_cb).Run(drop_event, output_drag_op);
+  return output_drag_op;
 }
 
 DropHelper::DropCallback DropHelper::GetDropCallback(
